### PR TITLE
Plane: Soaring upgrades

### PR DESCRIPTION
**Pull request by samuelctabor**
Various changes to the soaring code to make it more robust, flexible and provide more information to the user.

 - Change of coordinate system to NED for more intuitive logging
 - Much better estimation of air vertical speed
 - Exit soaring when properly lined up to next destination
 - Watchdog on actual climb rate to make sure to exit if not climbing
 - Add an autotest script
 - Various others

Already tested by @robustini and myself.

### CODE REVIEW

**Comment by robustini** [comment]
After countless tests I believe the new SOAR is mature enough to be merged in the master.
I've an infinite number of df logs that prove its perfect functioning.

immagine

**Comment by robustini** [comment]
Yesterday I also carried out a lot of tests with my ASW28 (4 meters wingspan, 6 Kg MTOM), with amazing results, I'm really satisfied with our work!
It's time to merge!

immagine

**Comment by samuelctabor** [comment]
Replaces

**Comment by WillyZehnder** [comment]
I've tested that #12210 several times and it's working very sophisticated. 
My test on last Saturday 26.11.19 lastet 1h 40m - then I was exhausted and landed (see  https://www.dropbox.com/sh/2mjss4mj55tmux3/AAAJidJyqd-Y9Pj3V8vNeQPZa?dl=0)

Especially that change in AP_TECS.cpp is a breakthrough for sailplanes, because now it's possible to use AdruPlane and SOAR on a pure sailplane (no motor) with no need to use a speed-sensor. So I bet many Scale-Saleplanes will be equipped with ArduPilot in the next year.

**Comment by SquashSquishSquaw** [comment]
These soaring upgrades are a big step forward in making ArduPlane the software of choice for glider pilots!

I'm looking forward to seeing them in the next stable version, hopefully together with the latest bugfixes by Alex/Yaapu ( https://github.com/ArduPilot/ardupilot/issues/12245 )!

**Comment by samuelctabor** [comment]
Thanks for the testing folks.

Rebased on latest master

**Comment by SquashSquishSquaw** [comment]
> These soaring upgrades are a big step forward in making ArduPlane the software of choice for glider pilots!
> 
> I'm looking forward to seeing them in the next stable version, hopefully together with the latest bugfixes by Alex/Yaapu ( #12245 )!

Has anything happened, in this respect? Could somebody please merge it? It would be sooo nice to have this in the stable version! Many thanks!

**Comment by auturgy** [comment]
Devcall tag added just to get this in the conversation before it bit rots.

File: ArduPlane/mode_loiter.cpp [hunk 1]
```diff
@@ -25,3 +27,56 @@ void ModeLoiter::update()
...
+         */
+        return true;
+    }
+
+    // Bearing in centi-degrees
+    const int32_t bearing_cd = plane.current_loc.get_bearing_to(targetLoc);
+    return isHeadingLinedUp(bearing_cd);
+}
+
+bool ModeLoiter::isHeadingLinedUp(const float bearing)
```
Comment (WickedShell) [inline ArduPlane/mode_loiter.cpp]: the float/int overload here is going to be much to tight, and cause people to get this wrong. The float one should just be removed, and the other one should be renamed possibly as far as `isHeadingLinedUp_cd` or something, to make it clearer.

As an aside I can't find anything that calls this version, can we just remove it?

File: ArduPlane/mode_loiter.cpp [hunk 2]
```diff
@@ -25,3 +27,56 @@ void ModeLoiter::update()
     plane.calc_throttle();
 }
 
+bool ModeLoiter::isHeadingLinedUp(const Location loiterCenterLoc, const Location targetLoc)
```
Comment (tridge) [inline ArduPlane/mode_loiter.cpp]: needs a comment on the function

File: ArduPlane/mode_loiter.cpp (hunk 1 above)
Comment (tridge) [inline ArduPlane/mode_loiter.cpp]: needs comment and change name

File: ArduPlane/mode_loiter.cpp [hunk 3]
```diff
@@ -25,3 +27,56 @@ void ModeLoiter::update()
...
+    return isHeadingLinedUp(bearing_cd);
+}
+
+bool ModeLoiter::isHeadingLinedUp(const float bearing)
+{
+    const int32_t bearing_cd = bearing*100;
+    return isHeadingLinedUp(bearing_cd);
+}
+
+bool ModeLoiter::isHeadingLinedUp(const int32_t bearing_cd)
```
Comment (tridge) [inline ArduPlane/mode_loiter.cpp]: same function name with int32??? please no

File: ArduPlane/servos.cpp [hunk 4]
```diff
@@ -470,16 +470,6 @@ void Plane::set_servos_controlled(void)
         SRV_Channels::set_output_scaled(SRV_Channel::k_throttle, 
             constrain_int16(quadplane.forward_throttle_pct(), min_throttle, max_throttle));
     }
-
-#if SOARING_ENABLED == ENABLED
-    // suppress throttle when soaring is active
```
Comment (tridge) [inline ArduPlane/servos.cpp]: did you mean to remove this?

File: ArduPlane/commands_logic.cpp [hunk 5]
```diff
@@ -1074,47 +1074,9 @@ bool Plane::verify_loiter_heading(bool init)
...
-    int32_t bearing_cd = current_loc.get_bearing_to(next_nav_cmd.content.location);
-
-    // get current heading.
-    int32_t heading_cd = gps.ground_course_cd();
-
-    int32_t heading_err_cd = wrap_180_cd(bearing_cd - heading_cd);
-
     if (init) {
-        loiter.sum_cd = 0;
+        plane.loiter.sum_cd = 0;
```
Comment (tridge) [inline ArduPlane/commands_logic.cpp]: why?

File: ArduPlane/commands_logic.cpp [hunk 6]
```diff
@@ -1074,47 +1074,9 @@ bool Plane::verify_loiter_heading(bool init)
...
-        // Want to head in a straight line from _here_ to the next waypoint instead of center of loiter wp
-
-        // 0 to xtrack from center of waypoint, 1 to xtrack from tangent exit location
-        if (next_WP_loc.loiter_xtrack) {
-            next_WP_loc = current_loc;
-        }
-        return true;
-    }
-    return false;
+    return plane.mode_loiter.isHeadingLinedUp(next_WP_loc, next_nav_cmd.content.location);
```
Comment (tridge) [inline ArduPlane/commands_logic.cpp]: needs review by @WickedShell

File: ArduPlane/mode_loiter.cpp (hunk 1 above)
Comment (samuelctabor) [reply ArduPlane/mode_loiter.cpp]: Removed this version and renamed the other function as you suggested.

File: ArduPlane/mode_loiter.cpp (hunk 2 above)
Comment (samuelctabor) [reply ArduPlane/mode_loiter.cpp]: Added comment.

File: ArduPlane/mode_loiter.cpp (hunk 1 above)
Comment (samuelctabor) [reply ArduPlane/mode_loiter.cpp]: Removed this version of the function.

File: ArduPlane/mode_loiter.cpp (hunk 3 above)
Comment (samuelctabor) [reply ArduPlane/mode_loiter.cpp]: Renamed function

File: ArduPlane/servos.cpp (hunk 4 above)
Comment (samuelctabor) [reply ArduPlane/servos.cpp]: Yes. This is now handled by setting a gliding_requested flag in the TECS, which will then output zero throttle. So this override is no longer required

File: ArduPlane/commands_logic.cpp (hunk 5 above)
Comment (samuelctabor) [reply ArduPlane/commands_logic.cpp]: Not sure - I assume this is referencing the same variable? If so I'll just revert this one

**Comment by samuelctabor** [comment]
I've created a PR against the wiki [here](https://github.com/ArduPilot/ardupilot_wiki/pull/2605) to document the new changes.

**Comment by samuelctabor** [comment]
Could somebody add the dev call topic to get another look at this? I believe I've resolved the comments from the first round. The failing CI test is due to some mew parameters not showing up in the generated xml, I'll have a go at resolving this.

**Comment by magicrub** [comment]
@samuelctabor  there's currently a merge conflict. Needs a rebase.

**Comment by peterbarker** [comment]
@samuelctabor What's the rationale behind adding another complete test suite for the soaring code?  AFAICS your `fly_mission` function could become a `fly_soaring` test within ArduPlane and do away with a rather large amount of boilerplate.

I've looked at the current autotest conflicts and they look straight-foward to resolve.

File: libraries/AP_Soaring/AP_Soaring.h [hunk 7]
```diff
@@ -65,34 +71,61 @@ class SoaringController {
     AP_Float alt_max;
     AP_Float alt_min;
     AP_Float alt_cutoff;
+    AP_Float max_drift;
 
 public:
     SoaringController(AP_AHRS &ahrs, AP_SpdHgtControl &spdHgt, const AP_Vehicle::FixedWing &parms);
 
+    typedef enum LoiterStatus {
```
Comment (peterbarker) [inline libraries/AP_Soaring/AP_Soaring.h:79]: We prefer `enum class` now-adays to avoid namespace conflicts.

**Comment by samuelctabor** [comment]
@peterbarker I did it this way to allow the plane-soaring frame to be used, so that appropriate aircraft mass and drag coefficients are applied in the simulation.

It looks like I could reduce boilerplate by defining AutoTestSoaring as a subclass of AutoTestPlane (in the same way that AutoTestHeli is a subclass of AutoTestCopter). Happy to refactor along those lines if that would be the preferred approach.

**Comment by peterbarker** [comment]
Any reason not to just make it a test in the Plane suite?

The chief advantage of doing that is so that we don't need to plumb it into places like `build_ci.sh` so these new tests prevent regressions *before* things get merged into master :-)

The Heli tests are separate because it's a different build target.

I'm not a great fan of the quadplane tests being split out as they are.  AFAICS we really ought to have one foo.py per binary produced.

**Comment by samuelctabor** [comment]
Makes sense! I can definitely do that if there is a way of specifying that the soaring test within the Plane suite requires a specific frame type - I couldn't find an example of this in the existing tests.

**Comment by peterbarker** [comment]
> Makes sense! I can definitely do that if there is a way of specifying that the soaring test within the Plane suite requires a specific frame type - I couldn't find an example of this in the existing tests.

Ah!  OK, that would be trailblazing somewhat.  Plan B.  Let's just get this in as-is and I'll clean it up later ;-)  (unless you'd really like to take it on)

Patches would be around `customise_sitl_commandline`, I think - passing in a frame and defaults file as new arguments to that method.

File: libraries/AP_Soaring/AP_Soaring.h (hunk 7 above)
Comment (samuelctabor) [reply libraries/AP_Soaring/AP_Soaring.h]: Refactored to use enum class

**Comment by samuelctabor** [comment]
> > Makes sense! I can definitely do that if there is a way of specifying that the soaring test within the Plane suite requires a specific frame type - I couldn't find an example of this in the existing tests.
> 
> Ah! OK, that would be trailblazing somewhat. Plan B. Let's just get this in as-is and I'll clean it up later ;-) (unless you'd really like to take it on)
> 
> Patches would be around `customise_sitl_commandline`, I think - passing in a frame and defaults file as new arguments to that method.

I've made the soaring test a subclass of Plane for now - at least that has reduced the boilerplate, but doesn't solve the CI issue. I'll have a look at making it a test within Plane as a separate PR.

**Comment by tridge** [comment]
remaining things needed:
 - checks from MdB for the changes to command logic for loiter
 - check under valgrind both with and without soaring enabled

**Comment by samuelctabor** [comment]
> remaining things needed:
> 
> * checks from MdB for the changes to command logic for loiter
> * check under valgrind both with and without soaring enabled

Thanks for the review - I ran valgrind against test.Soaring and against test.Plane and I have empty files named ``arduplane-plane-soaring-valgrind.log`` and ``arduplane-plane-elevrev-valgrind.log`` in the autotest directory - I gather that's a good thing :)

**Comment by tridge** [comment]
@samuelctabor thanks for your patience on this PR !

**Comment by WillyZehnder** [comment]
@samuelctabor Bril work Sam - Thank you very much.

**Pull request merged**